File: lib/kmeans.rb
# Classifies by K-Means method
# http://en.wikipedia.org/wiki/K-means_clustering
class KMeans

  # Clusters points into k clusters by euclidean distance.
  def cluster(k, points)
    centroids = pick_initial_centroids(k, points)
    while true
      newp = points.group_by { |p| closest_centroid(centroids, p) }
      newc = k.times.collect { |i| centroid(newp[i] || []) }
      return newp if newc == centroids
      centroids = newc
    end
  end

  private
  def closest_centroid(centroids, p)
    distances = centroids.collect { |c| distance(p, c) }
    closest = centroids.size.times.sort do |a,b| 
      distances[a] - distances[b]
    end
    closest.first
  end

  private
  def pick_initial_centroids(k, points)
    points.shuffle[0, k]
  end

  private
  def distance(a, b)
    dist = a.zip(b).collect { |i| (i.first - i.last) ** 2 }
    d = dist.inject(0) { |a,b| a + b }
  end

  private
  def centroid(points)
    return nil if points.empty?
    size = points.first.size
    sums = Array.new(size, 0.0)
    size.times do |i|
      points.each do |p|
        sums[i] += p[i]
      end
      sums[i] = sums[i] / points.size
    end
    sums
  end
end

File: brains/kmeans.py
import random

class KMeans(object):
    "Clusters data points by k-means method."

    def cluster(s, k, points):
        "Clusters by euclidean distance. Returns clustered points as a list."
        assert k > 0 and points is not None
        centroids = s._pick_initial_centroids(k, points)
        while True:
            newp = [list() for i in range(k)]
            for p in points:
                closest = s._closest_centroid(p, centroids)
                newp[closest].append(p)
            newc = [s._centroid(newp[i]) or centroids[i] for i in range(k)]
            if newc == centroids:
                return newp
            centroids = newc

    def _closest_centroid(s, p, centroids):
        assert p and centroids
        distances = [s._distance(p, c) for c in centroids]
        closest = sorted(range(len(distances)), lambda a, b: int(distances[a] - distances[b]))
        return closest[0]

    def _pick_initial_centroids(s, k, points):
        assert k > 0 and points
        random.shuffle(points)
        return points[:k]

    def _distance(s, a, b):
        assert a and b
        dist = 0
        for k in a.keys():
            dist += (a[k] - b[k]) ** 2
        return dist

    def _centroid(s, points):
        if not points:
            return None
        sums = {}
        for k in points[0].keys():
            for p in points:
                sums.setdefault(k, 0)
                sums[k] += p[k]
            sums[k] = sums[k] / len(points)
        return sums
